[AIRFLOW-1340] Add S3 to Redshift transfer operator

sekikn opened the pull request:
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1340


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Currently RedshiftToS3Transfer (UNLOAD) exists but the opposite doesn't.
This PR adds COPY operation as S3ToRedshiftTransfer.


### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Newly added tests.operators.test_s3_to_redshift_operator.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

artwr commented on airflow/operators/s3_to_redshift_operator.py:
```diff
@@ -0,0 +1,86 @@
+
+        copy_query = """
+                        COPY {0}.{1}
+                        FROM 's3://{2}/{3}/{1}_'
```
Can I ask what the purpose of the trailing underscore is?

sekikn replied:
@artwr Thanks for the review and sorry for the late reply.
It corresponds to [the opposite operation](/apache/incubator-airflow/blob/b918a4976a641a3c944ef4ad4323fe25bd219ea0/airflow/operators/redshift_to_s3_operator.py#L96) (UNLOAD in RedshiftToS3Transfer), which outputs files named tablename + underscore + postfix that Redshift appends. For example, when RedshiftToS3Transfer unloads the table called foo, it will dump it to the files named foo_xxxx_part_xx. So `COPY schema.foo FROM 's3://bucket/prefix/foo_` loads these files to S3, since their paths are forwardly matched with the path specified in the FROM clause.

But I just noticed that in the case of using S3ToRedshiftTransfer solely, it's an unnecessary limitation that the input files must contain underscore in there name. And without that limitation, S3ToRedshiftTransfer works with the files dumped by RedshiftToS3Transfer. So I'll remove the underscore in the current PR.

Furthermore, these operators should be able to specify the table name and input file name independently. I'll file it as a new JIRA issue.

artwr replied:
👍

Fokko commented on tests/operators/test_s3_to_redshift_operator.py:
```diff
@@ -0,0 +1,65 @@
+            """.format(schema, table, s3_bucket, s3_key,
+                       access_key, secret_key, copy_options)
+
+        def _trim(s):
```
Just curious, why is this `_trim` needed?

sekikn replied:
Without that, the following equality assertion can fail due to the number of spaces. For example, different indentation such as

```
    """
    COPY ...
    FROM ...
    """
```

and


```
    """
        COPY ...
        FROM ...
    """
```

will generate different numbers of spaces between `COPY ...` and `FROM`.

Fokko replied:
Makes sense, thanks for the explanation

Fokko commented on tests/operators/test_s3_to_redshift_operator.py:
```diff
@@ -0,0 +1,65 @@
+            with
+            credentials 'aws_access_key_id={4};aws_secret_access_key={5}'
+            {6};
+            """.format(schema, table, s3_bucket, s3_key,
```
Also here, could we use the dict string formatting to make it easier to read? Thanks!

Fokko reviewed (changes requested):
One minor change, apart from that, LGTM

sekikn committed:
[AIRFLOW-1340] Add S3 to Redshift transfer operator

Currently RedshiftToS3Transfer (UNLOAD) exists but the opposite doesn't.
This PR adds COPY operation as S3ToRedshiftTransfer.

Fokko reviewed (approved):
Thanks @sekikn. Looks good, merging.
